spark2/src/test/java/org/apache/iceberg/spark/source/TestStructuredStreamingRead.java

```diff
@@ -0,0 +1,489 @@
+    Collections.reverse(snapshotIds);
+
+    // Default max size per batch, this will consume all the data of this table.
+    DataSourceOptions options = new DataSourceOptions(ImmutableMap.of(
+        "path", tableLocation.toString(),
+        "checkpointLocation", checkpoint.toString()));
+    StreamingReader streamingReader = (StreamingReader) source.createMicroBatchReader(
+        Optional.empty(), checkpoint.toString(), options);
+    streamingReader.setOffsetRange(Optional.empty(), Optional.empty());
+
+    StreamingOffset start = (StreamingOffset) streamingReader.getStartOffset();
+    Assert.assertEquals("Start offset's snapshot id should be 1st snapshot id",
+        snapshotIds.get(0).longValue(), start.snapshotId());
+    Assert.assertEquals("Start offset's index should be the start index of 1st snapshot", 0, start.index());
+    Assert.assertTrue("Start offset's snapshot id should be a starting snapshot id", start.isStartingSnapshotId());
+    Assert.assertFalse("Start offset should not be the last index of 1st snapshot", start.isLastIndexOfSnapshot());
+
+    StreamingOffset end = (StreamingOffset) streamingReader.getEndOffset();
+    Assert.assertEquals("End offset's snapshot should be the last snapshot id",
+        snapshotIds.get(3).longValue(), end.snapshotId());
+    Assert.assertEquals("End offset's index should be the last index", 1, end.index());
+    Assert.assertFalse("End offset's snapshot id should not a starting snapshot id", end.isStartingSnapshotId());
+    Assert.assertTrue("End offset should be the last index of 3rd snapshot", end.isLastIndexOfSnapshot());
+
+    streamingReader.setOffsetRange(Optional.of(end), Optional.empty());
+    StreamingOffset end1 = (StreamingOffset) streamingReader.getEndOffset();
+    Assert.assertEquals("End offset should be same to start offset since there's no more batches to consume",
+        end1, end);
+  }
+
+  @SuppressWarnings("unchecked")
+  @Test
+  public void testGetOffsetWithRateLimit1000() throws IOException {
+    File checkpoint = Files.createTempDirectory(parent, "checkpoint").toFile();
+
+    IcebergSource source = new IcebergSource();
+    List<Long> snapshotIds = SnapshotUtil.currentAncestors(table);
+    Collections.reverse(snapshotIds);
+
+    // Max size to 1000, this will generate two MicroBatches per consuming.
+    DataSourceOptions options = new DataSourceOptions(ImmutableMap.of(
+        "path", tableLocation.toString(),
+        "checkpointLocation", checkpoint.toString(),
+        "max-size-per-batch", "1000"));
+    StreamingReader streamingReader = (StreamingReader) source.createMicroBatchReader(
+        Optional.empty(), checkpoint.toString(), options);
+
+    streamingReader.setOffsetRange(Optional.empty(), Optional.empty());
+    StreamingOffset start = (StreamingOffset) streamingReader.getStartOffset();
+    Assert.assertEquals("Start offset's snapshot id should be 1st snapshot id",
```

Comment:
I think we repeat this set of assertions a lot, we can probably wrap them in a generic private method for checking Offests